transcom/mymove
# Add payment request number to each payment request

## Description

This PR adds a payment request number to each payment request.  A payment request number is meant to be a unique human-readable identifier of the form `{MTO#}-{Incrementing payment request #}` where `MTO#` is the `reference_id` of the `move_task_order`.  In addition, the payment request number is returned in the payment request payload now for both the prime and GHC APIs.

## Reviewer Notes

What do we want to call this non-UUID identifier?  For now, it's a `payment request number`, but we may want a better name.  The story uses the term `payment request ID`, but we tend to use `ID` for UUIDs (which this isn't).

I had to do some locking when creating a payment request in order to ensure sequential numbers were obtained and we didn't have any clashes.  I use the `FOR NO KEY UPDATE` lock to minimize what gets locked for better concurrency.

Because payment requests may already have been created in some environments, the migration has a SQL function to add appropriate payment request numbers before setting the field to `not null`.

**I plan to deploy this to experimental before merging**, but wanted to get a general review first.

## Setup

1. If this is your first time connecting via Mutual TLS, make sure `primelocal` is set in your `/etc/hosts` file ([reference](https://github.com/transcom/mymove#setup-hosts))
1. `make server_build`
1. `make db_dev_e2e_populate`
1. `make server_run`

### To test with `curl`, paste this command into a terminal window:

```
curl -X POST \
  -k \
  --cert ./config/tls/devlocal-mtls.cer  \
  --key ./config/tls/devlocal-mtls.key \
  https://primelocal:9443/prime/v1/payment-requests \
  -H 'Content-Type: application/json' \
  -d '{"isFinal": false, "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e", "serviceItems": [{"id": "9db1bf43-0964-44ff-8384-3297951f6781", "params": [{"key": "WeightActual", "value": "1234"}, {"key": "RequestedPickupDate", "value": "2019-12-16"}]}, {"id": "d886431c-c357-46b7-a084-a0c85dd496d3", "params": [{"key": "WeightActual", "value": "5678"}]}]}'
```

### To test with Postman:

1. Make sure your client cert is set up: Go to the wrench icon, then `Settings`, then `Certificates`.  Set the CRT file and KEY file to match the paths in the `curl` command above.  Use `primelocal:9443` for the host and port.
1. In `Settings`, under `General`, turn off `SSL certificate verification`.
1. Create a new `POST` request to `https://primelocal:9443/prime/v1/payment-requests`
1. In the `Body` tab, click the `raw` radio button, and paste in this JSON: `{"isFinal": false, "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e", "serviceItems": [{"id": "9db1bf43-0964-44ff-8384-3297951f6781", "params": [{"key": "WeightActual", "value": "1234"}, {"key": "RequestedPickupDate", "value": "2019-12-16"}]}, {"id": "d886431c-c357-46b7-a084-a0c85dd496d3", "params": [{"key": "WeightActual", "value": "5678"}]}]}`
1. Change the drop-down to the right of the radio buttons to `JSON` (the header should then automatically set `Content-Type` to `application/json`
1. Hit the `Send` button

## Code Review Verification Steps

* Any new migrations/schema changes:

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-1147) for this change
